File: demo/test/find-month-calendar.test.js
import { expect } from 'chai';
import Month from "../src/find-month-calendar.js"

describe('Month Calendar Tests', () => {
    let month;

    beforeEach(() => {
        month = new Month();
    });

    it('should parse date correctly', () => {
        const date = '12/2021';
        const parsedDate = month.parseDate(date);
        expect(parsedDate).to.deep.equal({ month: 12, year: 2021 });
    });

    it('should throw error for invalid date format', () => {
        const date = '12-2021';
        expect(() => month.parseDate(date)).to.throw('Improper string encoding');
    });

    it('should identify leap years correctly', () => {
        expect(month.isLeapYear(1900)).to.be.false;
        expect(month.isLeapYear(2000)).to.be.true;
        expect(month.isLeapYear(2020)).to.be.true;
    });

    it('should calculate day difference correctly', () => {
        const startDate = { month: 1, year: 1900 };
        const endDate = { month: 1, year: 1901 };
        const diff = month.getDayDiff(startDate, endDate);
        expect(diff).to.equal(365);
    });

    it('should correctly identify if start date is greater than end date', () => {
        const startDate = { month: 12, year: 2021 };
        const endDate = { month: 1, year: 2021 };
        const result = month.isGreater(startDate, endDate);
        expect(result).to.be.true;
    });

    it('should correctly identify if start date is not greater than end date', () => {
        const startDate = { month: 1, year: 2021 };
        const endDate = { month: 12, year: 2021 };
        const result = month.isGreater(startDate, endDate);
        expect(result).to.be.false;
    });
});

File: python/src/find_month_calendar_test.py
import unittest
from find_month_calendar import parse_date, is_leap_year, get_day_diff, is_greater 

class TestMonthCalendar(unittest.TestCase):

    def test_parse_date_correctly(self):
        date = '12/2021'
        parsed_date = parse_date(date)
        self.assertEqual(parsed_date, {'month': 12, 'year': 2021})
        
    def test_identify_leap_years_correctly(self):
        self.assertFalse(is_leap_year(1900))
        self.assertTrue(is_leap_year(2000))
        self.assertTrue(is_leap_year(2020))

    def test_calculate_day_difference_correctly(self):
        start_date = {'month': 1, 'year': 1900}
        end_date = {'month': 1, 'year': 1901}
        diff = get_day_diff(start_date, end_date)
        self.assertEqual(diff, 365)

    def test_identify_start_date_greater_than_end_date(self):
        start_date = {'month': 12, 'year': 2021}
        end_date = {'month': 1, 'year': 2021}
        result = is_greater(start_date, end_date)
        self.assertTrue(result)

    def test_identify_start_date_not_greater_than_end_date(self):
        start_date = {'month': 1, 'year': 2021}
        end_date = {'month': 12, 'year': 2021}
        result = is_greater(start_date, end_date)
        self.assertFalse(result)

if __name__ == '__main__':
    unittest.main()